### PR TITLE
[ImportVerilog] Fix Increment and support Conditional, Set membership, Case, Parameter, Generate

### PR DESCRIPTION
**Pull request by angelzzzzz**
### [[ImportVerilog] Fix Increment operator](https://github.com/fabianschuiki/circt/pull/32/commits/d93bf035aa6cab5e473b2e2c18c8b7c240133a46)

Fixed the issue of Postincrement and Postdecrement not being implemented correctly.
Before fixing the issue, the results of `a = b++;` are as follows:
```
 %2 = moore.constant 1 : !moore.int
 %3 = moore.add %b, %2 : !moore.int
 moore.mir.bpassign %b, %3 : !moore.int
 moore.mir.bpassign %a, %b : !moore.int
``` 
And now, the results of `a = b++;` are as follows:
```
 %2 = moore.constant 1 : !moore.int
 %3 = moore.add %b, %2 : !moore.int
 %preValue = moore.variable %b : !moore.int
 moore.mir.bpassign %b, %3 : !moore.int
 moore.mir.bpassign %a, %preValue : !moore.int
``` 
Here, I use `%preValue` to store the value of `%b` before self increasing.

### [[ImportVerilog] Add Conditional operator](https://github.com/fabianschuiki/circt/pull/32/commits/8483b7f5bbc9c3689bb9b1376022715dc5c90073)
Use `scf.if` to translate Conditional operator, for example, the results of ` c = a > b ? a : b; ` are as follows:
```
%0 = moore.gt %a, %b : !moore.int -> !moore.bit
%1 = moore.constant false : !moore.bit
%2 = moore.mir.ne %0, %1 : (!moore.bit) -> i1
%3 = scf.if %2 -> (!moore.int) {
  scf.yield %a : !moore.int
} else {
  scf.yield %b : !moore.int
}
moore.mir.bpassign %c, %3 : !moore.int
``` 

### [[ImportVerilog] Add Set membership operator](https://github.com/fabianschuiki/circt/pull/32/commits/291938496bb2eecc8837cab54b3ddf3ca310d7d9)
Use `scf.if` to translate Set membership operator -- `Inside` operator, for example, the results of ` c = a inside {a,b}; ` are as follows:
```
%0 = moore.mir.eq %a, %a : (!moore.int) -> i1
%1 = scf.if %0 -> (!moore.logic) {
  %4 = moore.constant true : !moore.logic
  scf.yield %4 : !moore.logic
} else {
  %4 = moore.mir.eq %a, %b : (!moore.int) -> i1
  %5 = scf.if %4 -> (!moore.logic) {
    %6 = moore.constant true : !moore.logic
    scf.yield %6 : !moore.logic
  } else {
    %6 = moore.constant false : !moore.logic
    scf.yield %6 : !moore.logic
  }
  scf.yield %5 : !moore.logic
}
%2 = moore.conversion %1 : !moore.logic -> !moore.packed<range<logic, 31:0>>
%3 = moore.conversion %2 : !moore.packed<range<logic, 31:0>> -> !moore.int
moore.mir.bpassign %c, %3 : !moore.int
``` 
### [[ImportVerilog] Add Case statement](https://github.com/fabianschuiki/circt/pull/32/commits/79117325397ca070641a15b728268eb9860371db)
Use `scf.if` to translate Case statement, for example, translate the code:
```
case (a)
  b: ;
  default ;
endcase
```
results:
```
%0 = moore.mir.eq %a, %b : (!moore.int) -> i1
scf.if %0 {
} else {
}
```
### [[ImportVerilog] Support Parameter and Localparam](https://github.com/fabianschuiki/circt/pull/32/commits/86b679b05d88b76c59f917e1304f0039d2ed57b9)
Expand ConstantOp and support Parameter by adding name after value, for example, translate the code:
```
const int a = 1;
parameter b = 2;
```
results:
```
%0 = moore.constant 1 : !moore.int
%a = moore.variable %0 : !moore.int
%1 = moore.constant 2 "b" : !moore.packed<range<logic<signed>, 31:0>>
```
We use name to distinguish between `const` and `parameter`.
### [[ImportVerilog] Support Generate constructs](https://github.com/fabianschuiki/circt/pull/32/commits/db0446fa439341706201e47eac40093355507c83)
* In order to parse the member of GenerateBlock, I extract the parsing loop of InstanceBody and put it into `walkMember()` function in `Structure.cpp`. 
* Because `genvar` only exists in declaration and the variable exists as `parameter` in GenerateBlock, `genvar` can be skipped.
* Fix the issue that `parameter` should be added into SymbolTable.

### CODE REVIEW

File: include/circt/Dialect/Moore/MooreOps.td [hunk 1]
```diff
@@ -188,13 +188,14 @@ def EventControlOp : MooreOp<"event", [
...
+  let arguments = (ins APIntAttr:$value,OptionalAttr<StrAttr>:$name);
   let results = (outs SimpleBitVectorType:$result);
   let hasCustomAssemblyFormat = 1;
   let hasVerifier = 1;
   let builders = [
     OpBuilder<(ins "Type":$type, "const APInt &":$value)>,
     OpBuilder<(ins "Type":$type, "int64_t":$value)>,
+    OpBuilder<(ins "Type":$type, "int64_t":$value, "StringRef":$name)>,
   ];
 }
```
Comment (fabianschuiki) [inline include/circt/Dialect/Moore/MooreOps.td]: Constants shouldn't have names. Most ops in the IR should not have names, unless they explicitly represent a construct in the source language that is named, such as a port, variable, or net.

File: include/circt/Dialect/Moore/MIRExpressions.td [hunk 2]
```diff
@@ -133,14 +133,15 @@ def ShrOp : ShiftOp<"shr"> {
...
 ]> {
   let arguments = (ins SimpleBitVectorType:$lhs,
                        SimpleBitVectorType:$rhs,
                        UnitAttr:$mode);
   let results = (outs I1:$result);
   let assemblyFormat = [{
-    ( `case` $mode^ )? $lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs)
+    ( `case` $mode^ )? $lhs `,` $rhs attr-dict `:` functional-type($lhs,$result)
   }];
 }
```
Comment (fabianschuiki) [inline include/circt/Dialect/Moore/MIRExpressions.td]: I think I forgot to delete the `EqualOp` here. There's a replacement op in `MooreOps.td` (or similar), I think called `EqOp`. You should be able to just delete this.

File: lib/Conversion/ImportVerilog/Expression.cpp [hunk 3]
```diff
@@ -114,6 +114,10 @@ struct ExprVisitor {
     auto postValue =
         isInc ? builder.create<moore::AddOp>(loc, preValue, one).getResult()
               : builder.create<moore::SubOp>(loc, preValue, one).getResult();
+    if (isPost) {
+      preValue = builder.create<moore::VariableOp>(loc, preValue.getType(),
+                                                   "preValue", arg);
+    }
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/Expression.cpp]: Thanks for catching this issue! I don't think you need a separate variable though. You should be able to just swap the order of the assignment and the add/sub op to implement the pre- and post-increment flavors. The variable op will not do what you expect: its initial value is only assigned once at the beginning of the simulation, and won't be updated on subsequent clock ticks. It also introduces a named value into the simulation which doesn't correspond to any named value in the source language, which we should avoid.

Comment (fabianschuiki) [reply lib/Conversion/ImportVerilog/Expression.cpp]: You'll probably want to fix the following broken IR:
```mlir
 %2 = moore.constant 1 : !moore.int
 %3 = moore.add %b, %2 : !moore.int
 moore.mir.bpassign %b, %3 : !moore.int
 moore.mir.bpassign %a, %b : !moore.int
```
by moving the assignment:
```mlir
 moore.mir.bpassign %a, %b : !moore.int
 %2 = moore.constant 1 : !moore.int
 %3 = moore.add %b, %2 : !moore.int
 moore.mir.bpassign %b, %3 : !moore.int
```

File: lib/Conversion/ImportVerilog/Expression.cpp [hunk 4]
```diff
@@ -267,6 +271,65 @@ struct ExprVisitor {
...
+  Value visit(const slang::ast::ConditionalExpression &expr) {
+    auto type = context.convertType(*expr.conditions.begin()->expr->type);
+    Value cond = convertToSimpleBitVector(
+        context.convertExpression(*expr.conditions.begin()->expr));
+    if (!cond)
+      return {};
+    if (!cond.getType().isa<mlir::IntegerType>()) {
+      auto zeroValue = builder.create<moore::ConstantOp>(loc, type, 0);
+      cond = builder.create<moore::InEqualityOp>(loc, cond, zeroValue);
+    }
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/Expression.cpp]: I think you want a boolean cast here instead. Check out the lowering of the `if` statement. I think that had to do a similar thing and that might be using the `bool_cast` op to capture this.

File: lib/Conversion/ImportVerilog/Expression.cpp [hunk 5]
```diff
@@ -267,6 +271,65 @@ struct ExprVisitor {
...
+
+  Value visit(const slang::ast::InsideExpression &expr) {
+    auto lhs = convertToSimpleBitVector(context.convertExpression(expr.left()));
+    auto rangeList = expr.rangeList();
+    if (!lhs || rangeList.empty())
+      return {};
+    auto type = context.convertType(*expr.type);
+    return createInside(rangeList, 0, lhs, loc, type);
+  }
+
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/Expression.cpp]: Cool! Great to see the `inside` expression being supported :heart_eyes:. Could you make it non-recursive? Instead of calling `createInside` recursively, could you just iterate over the `rangeList` and build up a sequence of `ifOp`s?

File: lib/Conversion/ImportVerilog/Statement.cpp [hunk 6]
```diff
@@ -54,6 +54,55 @@ LogicalResult Context::visitConditionalStmt(
...
+    builder.setInsertionPoint(ifOp.elseYield());
+    if (exprIndex < (items[itemIndex].expressions.size() - 1)) {
+      return createCase(caseExpr, items, defaultCase, itemIndex, exprIndex + 1,
+                        loc, type);
+    }
+
+    return createCase(caseExpr, items, defaultCase, itemIndex + 1, 0, loc,
+                      type);
+  }
+  return success();
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/Statement.cpp]: Similar to the `inside` expression above: can you make this non-recursive and instead iterate over the `items` and build up a sequence of `ifOp`s?

File: test/Conversion/ImportVerilog/basic.sv [hunk 7]
```diff
@@ -180,6 +182,37 @@ module Expressions();
...
+    // CHECK:   } else {
+    // CHECK:     [[TMP7:%.+]] = moore.constant false : !moore.logic
+    // CHECK:     scf.yield [[TMP7]] : !moore.logic
+    // CHECK:   }
+    // CHECK:   scf.yield [[TMP6]] : !moore.logic
+    // CHECK: }
+    // CHECK: [[TMP3:%.+]] = moore.conversion [[TMP2]] : !moore.logic -> !moore.packed<range<logic, 31:0>>
+    // CHECK: [[TMP4:%.+]] = moore.conversion [[TMP3]] : !moore.packed<range<logic, 31:0>> -> !moore.int
+    // CHECK: moore.mir.bpassign %c, [[TMP4]] : !moore.int
+    c = a inside {a, b};
```
Comment (fabianschuiki) [inline test/Conversion/ImportVerilog/basic.sv]: Can you also test that `inside {a}` and `inside {a, b, a, b}` work as expected?

File: test/Conversion/ImportVerilog/basic.sv [hunk 8]
```diff
@@ -239,9 +272,62 @@ module Statements();
...
+  generate
+    // CHECK: [[TMP1:%.+]] = moore.constant 0 "i" : !moore.integer
+    // CHECK: [[TMP2:%.+]] = moore.conversion [[TMP1]] : !moore.integer -> !moore.int
+    // CHECK: %c = moore.variable [[TMP2]] : !moore.int
+    // CHECK: [[TMP3:%.+]] = moore.constant 1 "i" : !moore.integer
+    // CHECK: [[TMP4:%.+]] = moore.conversion [[TMP3]] : !moore.integer -> !moore.int
+    // CHECK: %c_0 = moore.variable name "c" [[TMP4]] : !moore.int
+    for(i=0;i<a;i=i+1)begin
+	    int c = i;
+	  end
```
Comment (fabianschuiki) [inline test/Conversion/ImportVerilog/basic.sv]: Can you introduce a separate op to assign a name to a compile-time constant value? Something like `%a = moore.named_constant %0`, more or less exactly how `moore.variable` and `moore.net` work? That will give us a clear indication in the IR as to what parts need to be fully constant, and it will decouple the naming of a constant value from defining the actual value itself. Later on if we need it, we could even add an indication to that op to represent whether it was a `parameter`, `localparam`, `specparam`, or `genvar`.

File: test/Conversion/ImportVerilog/errors.sv [hunk 9]
```diff
@@ -16,5 +16,5 @@ endmodule
 
 module Foo;
   // expected-warning @below {{unsupported construct ignored}}
-  genvar a;
+  specparam sp =8;
```
Comment (fabianschuiki) [inline test/Conversion/ImportVerilog/errors.sv]:
```suggestion
  specparam sp = 8;
```

File: test/Conversion/ImportVerilog/basic.sv (hunk 8 above)
Comment (fabianschuiki) [reply test/Conversion/ImportVerilog/basic.sv]: Love the idea of capturing these constants in the IR though :partying_face:!

**Review by fabianschuiki — CHANGES_REQUESTED** [review]

File: include/circt/Dialect/Moore/MIRExpressions.td (hunk 2 above)
Comment (terapines-osc-circt) [reply include/circt/Dialect/Moore/MIRExpressions.td]: I know it and agree to delete it. But now the condition of the scf::if needs I1 as its argument. I attempted to use the `SingleBitType` you defined, but `the type mismatch with I1` will be thrown.

File: lib/Conversion/ImportVerilog/Expression.cpp (hunk 4 above)
Comment (angelzzzzz) [reply lib/Conversion/ImportVerilog/Expression.cpp]: > I think you want a boolean cast here instead. Check out the lowering of the `if` statement. I think that had to do a similar thing and that might be using the `bool_cast` op to capture this. 

@fabianschuiki 
Thanks for your suggestion :heart: ! However, when using `bool_cast` op, It throws out an error that `'scf.if' op operand #0 must be 1-bit signless integer, but got '!moore.bit'`. The reason why it failed is that the `condition` of `scf.if` needs `I1` type rather than `bit` type. Given this, I add `ConversionOp` to convert `bit` type to `I1` type and apply it to `scf.if` op. The specific code is as follows：
```cpp
  Value visit(const slang::ast::ConditionalExpression &expr) {
    Value cond = convertToSimpleBitVector(
        context.convertExpression(*expr.conditions.begin()->expr));
    if (!cond)
      return {};

    if (auto condType = dyn_cast_or_null<moore::UnpackedType>(cond.getType())) {
      if (condType.isCastableToSimpleBitVector() && condType.getBitSize() != 1)
        cond = builder.create<moore::BoolCastOp>(loc, cond);
    }
    cond = builder.create<moore::ConversionOp>(loc, builder.getI1Type(), cond);

    auto ifOp = builder.create<mlir::scf::IfOp>(
        loc, cond,
        [&](OpBuilder &builder, Location loc) {
          builder.create<mlir::scf::YieldOp>(
              loc, context.convertExpression(expr.left()));
        },
        [&](OpBuilder &builder, Location loc) {
          builder.create<mlir::scf::YieldOp>(
              loc, context.convertExpression(expr.right()));
        });
    return ifOp.getResult(0);
  }
```
The results of `c = a ? a : b;` are as follows:
```
%0 = moore.bool_cast %a : !moore.int -> !moore.bit
%1 = moore.conversion %0 : !moore.bit -> i1
%2 = scf.if %1 -> (!moore.int) {
    scf.yield %a : !moore.int
  } else {
    scf.yield %b : !moore.int
  }
moore.mir.bpassign %c, %2 : !moore.int
```
I would like to know if it is OK. Looking forward to your reply! :heart:

Comment (fabianschuiki) [reply lib/Conversion/ImportVerilog/Expression.cpp]: Yeah I think this would definitely work! :smile:

**Comment by angelzzzzz** [comment]
Fix the issues mentioned above:
1. `parameter` type
    Support `parameter`, `localparam`, `specparam`, but `genvar` is ignored because it only appears in the declaration.
3. Delete the extra `EqualOp`
5. `post-increment` & `post-decrement` expressions
    By adding the function `createPostValue` in ImportVerilogInternals.h, we can store the value when coming across `a--` or `a++`, and print `a = a + 1` or `a = a - 1` after translating a statement or variable.
7. Add `bool_cast` op
8. `inside` op
    According to "IEEE Standard for SystemVerilog", which mentions that " The return value is effectively the or reduction of all the comparisons in the set with the expression on the left-hand side", I add `LogicalOrOp` to implement the selection of 0 and 1 instead of using `scf.if`.
10. `case` op
    Similar to the `inside` op, it uses `LogicalOrOp` to evaluate whether to execute the code of `default` case. 
12. Add tests of `inside` op
13. Redefine the logical operators.

I'd like to know if it's appropriate to add `LogicalOrOp` to the implementation of `inside` and `case`. :heart:

**Comment by angelzzzzz** [comment]
Similar to Expression.cpp, I add `convertToBool` function to Statement.cpp.

File: include/circt/Dialect/Moore/MooreOps.td [hunk 10]
```diff
@@ -256,6 +256,32 @@ def ConstantOp : MooreOp<"constant", [Pure]> {
...
+  let arguments = (ins StrAttr:$name, 
+                       NamedConstAttr:$kind, 
+                       I64Attr:$value);
+  let results = (outs UnpackedType:$result);
+  let assemblyFormat = [{
+    `` custom<ImplicitSSAName>($name) $kind $value attr-dict
+    `:` type($result)
+  }];
+}
+
```
Comment (fabianschuiki) [inline include/circt/Dialect/Moore/MooreOps.td:284]: Very nice, thanks!

File: lib/Conversion/ImportVerilog/ImportVerilogInternals.h [hunk 11]
```diff
@@ -56,12 +56,18 @@ struct Context {
...
+  LogicalResult walkMember(const slang::ast::Symbol &members);
   LogicalResult
   convertStatementBlock(const slang::ast::StatementBlockSymbol *stmt);
 
   // Convert a slang statement into an MLIR statement.
   LogicalResult convertStatement(const slang::ast::Statement *statement);
 
+  LogicalResult
+  visitConditionalStmt(const slang::ast::ConditionalStatement *conditionalStmt);
+
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/ImportVerilogInternals.h:68]: If I remember correctly, this was deleted on the `slang-frontend` branch. Might have gotten into your PR accidentally through merges/rebases.

File: lib/Conversion/ImportVerilog/Structure.cpp [hunk 12]
```diff
@@ -137,116 +137,177 @@ Context::convertModuleBody(const slang::ast::InstanceBodySymbol *module) {
...
+  }
+  return success();
+}
 
-    // Skip parameters. The AST is already monomorphized.
-    if (member.kind == slang::ast::SymbolKind::Parameter)
-      continue;
+LogicalResult Context::walkMember(const slang::ast::Symbol &member) {
+  LLVM_DEBUG(llvm::dbgs() << "- Handling " << slang::ast::toString(member.kind)
+                          << "\n");
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/Structure.cpp:148]: Nice factoring

File: test/Conversion/ImportVerilog/basic.sv [hunk 13]
```diff
@@ -283,10 +319,69 @@ module Statements();
...
+    // CHECK: %e = moore.variable [[TMP1]] : !moore.int
+	  case (a)
+      2:begin
+        int e = 2;
+      end
+      default:begin
+        int e = 3;
+      end
+    endcase
+  endgenerate
```
Comment (fabianschuiki) [inline test/Conversion/ImportVerilog/basic.sv:386]: It looks like there are a few tabs here instead of spaces. Can you change the indentation to all tabs?

File: test/lit.cfg.py [hunk 14]
```diff
@@ -58,8 +58,8 @@
 tools = [
     'arcilator', 'circt-as', 'circt-capi-ir-test', 'circt-capi-om-test',
     'circt-capi-firrtl-test', 'circt-capi-firtool-test', 'circt-dis',
-    'circt-opt', 'circt-reduce', 'circt-translate', 'firtool', 'hlstool',
-    'om-linker', 'ibistool'
+    'circt-opt', 'circt-reduce', 'circt-translate', 'circt-verilog', 'firtool',
+    'hlstool', 'om-linker', 'ibistool'
```
Comment (fabianschuiki) [inline test/lit.cfg.py:62]: Unrelated change. This got removed in `slang-frontend`. Probably came in accidentally through a merge/rebase.

File: test/circt-verilog/preprocess-errros.sv [hunk 15]
```diff
@@ -0,0 +1,4 @@
+// RUN: circt-verilog %s -E --verify-diagnostics
+
+// expected-error @below {{could not find or open include file}}
+`include "unknown.sv"
```
Comment (fabianschuiki) [inline test/circt-verilog/preprocess-errros.sv:4]: Old file. This got removed in `slang-frontend`. Probably came in accidentally through a merge/rebase.

File: test/Conversion/ImportVerilog/basic.sv [hunk 16]
```diff
@@ -283,10 +319,69 @@ module Statements();
...
+    // CHECK: }
+    // CHECK: [[TMP3:%.+]] = moore.not [[TMP1]] : !moore.bit
+    // CHECK: [[TMP4:%.+]] = moore.conversion [[TMP3]] : !moore.bit -> i1
+    // CHECK: scf.if [[TMP4]] {
+    // CHECK: }
+    case (a)
+      b: ;
+      default ;
+    endcase
+
```
Comment (fabianschuiki) [inline test/Conversion/ImportVerilog/basic.sv:341]: Nice! Can you add a second test for `case` with maybe 1 or 2 additional cases, to make sure the cases are chained properly. And also a third test for `case` with no `default`. That should exercise all code paths in your nice lowering above :slightly_smiling_face:

**Review by fabianschuiki — COMMENTED** [review]
Very nice! Just a few minor comments inline. Thanks a lot for doing all this work!

File: lib/Conversion/ImportVerilog/ImportVerilogInternals.h (hunk 11 above)
Comment (angelzzzzz) [reply lib/Conversion/ImportVerilog/ImportVerilogInternals.h:68]: Sorry, this is my mistake.

File: test/Conversion/ImportVerilog/basic.sv (hunk 13 above)
Comment (angelzzzzz) [reply test/Conversion/ImportVerilog/basic.sv:386]: Sure, I will fix it.

File: test/Conversion/ImportVerilog/basic.sv (hunk 16 above)
Comment (angelzzzzz) [reply test/Conversion/ImportVerilog/basic.sv:341]: Thank you very much for your advice :rose:, I will add more cases for this.
